### PR TITLE
Monte Carlo simulation model

### PR DESCRIPTION
**Pull request by linxin1994**
Model for Monte Carlo simulation

### CODE REVIEW

File: models/Monte_Carlo.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    delta_vy = np.sqrt(const.k*T/m)
+    delta_z = np.sqrt(const.k*T/(m*omega_L**2))
+    delta_vz = np.sqrt(const.k*T/m)
+    x_ensemble = np.random.normal(0, delta_x, N_ensemble)
+    y_ensemble = np.random.normal(0, delta_y, N_ensemble)
+    z_ensemble = np.random.normal(0, delta_z, N_ensemble)
+    vx_ensemble = np.random.normal(0, delta_vx, N_ensemble)
+    vy_ensemble = np.random.normal(0, delta_vy, N_ensemble)
+    vz_ensemble = np.random.normal(0, delta_vz, N_ensemble)
+
```
Comment (neutrinonerd3333) [inline models/Monte_Carlo.py:39]: !ETU !RFC: As an example of vectorization, this entire block could be condensed into four lines expressing your intent more clearly:

```python
delta_r = np.sqrt(const.k * T / (m * np.array([omega_H, omega_H, omega_L])**2))
delta_v = np.sqrt(const.k * T / m) * np.ones(3)
r_ensemble = np.random.normal(size=(N_ensemble, 3)) * delta_r
v_ensemble = np.random.normal(size=(N_ensemble, 3)) * delta_v
```

Whenever variables broken up into $x,y,z$ components here, it should be possible to rewrite the code more briefly and clearly in vector form.

File: models/Monte_Carlo.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    y_ensemble_t = np.array(y_ensemble_t)
+    z_ensemble_t = np.array(z_ensemble_t)
+    vx_ensemble_t = np.array(vx_ensemble_t)
+    vy_ensemble_t = np.array(vy_ensemble_t)
+    vz_ensemble_t = np.array(vz_ensemble_t)
+
+    return x_ensemble_t, vx_ensemble_t, y_ensemble_t, vy_ensemble_t, z_ensemble_t, vz_ensemble_t
+
+
+def recature_rate_3d(trap_temp, x_ensemble_t, vx_ensemble_t, y_ensemble, vy_ensemble, z_ensemble, vz_ensemble, t, m, omega_H, omega_L):
```
Comment (neutrinonerd3333) [inline models/Monte_Carlo.py:145]: !ETU: I think this whole function would vectorize down to just a few very clear lines :)

**Review by neutrinonerd3333 — CHANGES_REQUESTED** [review]
Sam and I just took a look through this. The physics looks good except for one point of confusion. We took a long time to check and understand everything; some small changes to coding style would help this going forward. The two main ones are: refactoring almost identical code ("DRY": don't repeat yourself) and taking advantage of `numpy` vectorizing to rewrite all the vector calculations more succinctly (vectorized operations almost always run faster too). Will accept after these changes are made :)

(key to comments: SFB = safe from bugs; ETU = easy to understand; RFC = ready for change)

File: models/Monte_Carlo.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    x_ensemble_t = []
+    y_ensemble_t = []
+    z_ensemble_t = []
+    vx_ensemble_t = []
+    vy_ensemble_t = []
+    vz_ensemble_t = []
+
+    for x, vx, y, vy, z, vz in zip(x_ensemble, vx_ensemble, y_ensemble, vy_ensemble, z_ensemble, vz_ensemble):
+        f0 = [x, vx, y, vy, z, vz]
+        sol = odeint(newton_3d, f0, t, args = (omega_iHO_H, omega_iHO_L,))
```
Comment (neutrinonerd3333) [inline models/Monte_Carlo.py:58]: Also -- I think that the equations of motion here are integrable, so you could just write $x(t) = x(0) \cosh \omega_x t + x'(0) \omega_x^{-1} \sinh \omega_x t$

File: models/Monte_Carlo.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    y_ensemble = np.random.normal(0, delta_y, N_ensemble)
+    z_ensemble = np.random.normal(0, delta_z, N_ensemble)
+    vx_ensemble = np.random.normal(0, delta_vx, N_ensemble)
+    vy_ensemble = np.random.normal(0, delta_vy, N_ensemble)
+    vz_ensemble = np.random.normal(0, delta_vz, N_ensemble)
+
+
+    v = wavelength2frequency(traping_laser)
+    v0 = wavelength2frequency(D2_line)
+    alpha = abs((v + v0)*(v-v0)/v**2)
```
[the remark this answers is not in the document]
Comment (npeard) [reply models/Monte_Carlo.py:43]: Add a citation comment in the code to: https://journals.aps.org/pra/pdf/10.1103/PhysRevA.97.053803

[the remark this answers is not in the document]
Comment (neutrinonerd3333) [reply models/Monte_Carlo.py:43]: Okay, Sam and I looked at this and we now understand what's going on. This should have a cite to the relevant part of the paper and a comment on what it is (`alpha` isn't very descriptive on its own 😅)

File: models/Monte_Carlo.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    x_ensemble = np.random.normal(0, delta_x, N_ensemble)
+    y_ensemble = np.random.normal(0, delta_y, N_ensemble)
+    z_ensemble = np.random.normal(0, delta_z, N_ensemble)
+    vx_ensemble = np.random.normal(0, delta_vx, N_ensemble)
+    vy_ensemble = np.random.normal(0, delta_vy, N_ensemble)
+    vz_ensemble = np.random.normal(0, delta_vz, N_ensemble)
+
+
+    v = wavelength2frequency(traping_laser)
+    v0 = wavelength2frequency(D2_line)
```
[the remark this answers is not in the document]
Comment (neutrinonerd3333) [reply models/Monte_Carlo.py:42]: Also a correctness issue -- in the referenced paper $\nu_0$ is the mean frequency of the $D$ lines.

**Review by npeard — APPROVED** [review]
changes look good

**Pull request merged**